### ISSUE TITLE
DI - Handle the "InjectFlags" in Injector_.get()

### ISSUE THREAD

**Issue by TomMenga** [angular/angular#31776]
# 🚀 feature request

### Relevant Package

 This feature request is for @angular/core

### Description
At the moment the `Injector_.get()` implementation doesn't handle the `InjectFlags`. 
 I'd like the feature so that I can use the component `Injector` like this:
```typescript
@Component(...)
export class SomeComponent {
    // Here I get an instance of Injector_
    constructor(injector: Injector) {
        this.myService = injector.get(SomeService, null, InjectFlags.SkipSelf);
    }
}
```

### Describe the solution you'd like
It looks like you already handle the `DeepFlags`.   [refs.ts:342](https://github.com/angular/angular/blob/da2880d7c495022405102f89b8c641b23516388d/packages/core/src/view/refs.ts#L342)  
You probably just need to map and pass the `InjectFlags` to the `resolveDeep()` function
```typescript
class Injector_ implements Injector {
  ...
  get(token: any, notFoundValue: any = Injector.THROW_IF_NOT_FOUND): any {
    ...
    return Services.resolveDep(
        this.view, this.elDef, allowPrivateServices,
        {flags: DepFlags.None, token, tokenKey: tokenKey(token)}, notFoundValue);
  }
}
```
I don't think I have enough knowledge of the project to submit a PR by myself, so I'm relying on you for this.  
Thanks in advance for your work

**Comment by pmilic021** [comment]
Whats the status on this one? If no one is taking care of it, I could go ahead and try to make a PR for a solution.

**Comment by stupidawesome** [comment]
`NodeInjector` isn't passing the flag here [di.ts#L628](https://github.com/angular/angular/blob/9.1.0-next.1/packages/core/src/render3/di.ts#L628)

```ts
export class NodeInjector implements Injector {
  constructor(
      private _tNode: TElementNode|TContainerNode|TElementContainerNode|null,
      private _lView: LView) {}

  get(token: any, notFoundValue?: any): any {
    return getOrCreateInjectable(this._tNode, this._lView, token, undefined, notFoundValue);
  }
}
```

**Comment by kpeterbauer** [comment]
I would call it a bug, not a feature: The provided Injector clearly does not behave as defined by the API documentation:

https://angular.io/api/core/Injector

**Comment by dimamarksman** [comment]
Looks like a bug, the type signature says us that there is third option for flag, but under the hood the implementation does not take this option into account.

**Comment by johncrim** [comment]
The likely workaround for this is going to be to use the `@SkipSelf()` decorator and not use the `Injector` API directly, but that's suboptimal in a lot of cases.

My use case is that I want to provide services at the directive level only if they're not already provided by an ancestor element, using the following:

Use case:
```ts
/** Trigger for opening a popup when an element (eg form element) is focused. */
@Directive({
  selector: '[ui-popup-onfocus]',
  providers: [
    provideIfMissing({
      provide: PopupController,
      factory: () => new PopupController()
    })
  ]
})
export class PopupOnFocusDirective {
```

Implementation, which doesn't work b/c it requires `InjectFlags.SkipSelf` (and `Optional`):
```ts
/**
 * Like a `FactoryProvider`, but the factory function is only called if the
 * dependency is not already provided by a parent injector.
 */
export interface IfMissingFactoryProvider<T> {
  /** The `Type<T>` or `InjectionToken<T>` to inject. */
  provide: Type<T> | InjectionToken<T>;
  /** A factory function to create a T */
  factory: (...args: any[]) => T;
  /** Injectable dependencies that are passed to the `factory` function. */
  deps?: any[];
}

/**
 * Injects an instance of type T if not already provided by a parent.
 *
 * @param provide A `Type<T>` or `InjectionToken<T>`
 * @param factory A factory function to create a T
 * @param deps Injectable dependencies that are passed to the `factory` function.
 */
export function provideIfMissing<T>(
  {
    provide,
    factory,
    deps
  }: IfMissingFactoryProvider<T>): FactoryProvider {

  let factoryDepth = 0;

  function ifMissingFactory(injector: Injector, ...args: any[]): T {
    if (factoryDepth === 0) {
      factoryDepth++; // Protect against recursion
      // tslint:disable-next-line: no-bitwise
      const t: T | null = injector.get<T>(provide, undefined, InjectFlags.Optional | InjectFlags.SkipSelf);
      factoryDepth--;
      if (t !== null) {
        return t;
      }
    }
    return factory(...args);
  }

  return {
    provide,
    useFactory: ifMissingFactory,
    deps: deps ? [ Injector, ...deps ] : [ Injector ]
  };
}
```

The workaround is for doing the same thing using `@SkipSelf` is quite a bit kludgier - will require constructor modification and injecting custom factories on each class in the chain, instead of just using a shared factory function.

**Comment by matthewjh** [comment]
Yes, need this. If the flags are not supported, the implementation needs to throw an error saying so

**Comment by johnc-ftl** [comment]
I've found that if I use `inject()` instead of `Injector.get()` I'm able to use the `InjectFlags`. At least it seems to work so far.

The limitation of this is that the `inject()` function only works in factory functions, you can't use it in regular code where `Injector` can be used. But with some refactoring use of `Injector` could usually be moved into factory functions.

**Comment by johncrim** [comment]
I've found that if I use `inject()` instead of `Injector.get()` I'm able to use the `InjectFlags` arg. At least it seems to work so far.

The limitation of this is that the `inject()` function only works in factory functions, you can't use it in regular code where `Injector` can be used. But with some refactoring use of `Injector` could usually be moved into factory functions.

**Comment by johncrim** [comment]
After some more testing, I've found that:

In Ivy w/ Angular 11.0.2:
* `inject(token, InjectFlags)` correctly uses the `InjectFlags`
* `Injector.get(token, defaultValue, InjectFlags)` throws away the `InjectFlags`
Note that `inject()` can only be used in factory functions.

In ViewEngine w/ Angular 11.0.2, in Jest tests:
* Calling `inject()` from a factory function throws an error: `inject() must be called from an injection context`
* `Injector.get(token, defaultValue, InjectFlags)` still throws away the `InjectFlags`

This inconsistency is a problem for me b/c we use Jest for testing, and jest-preset-angular is still compiling to ViewEngine (they're working on compiling using Ivy). My (hopefully temporary) workaround is going to be using karma for all the tests that require Ivy injection behavior.

**Comment by mrkeathley** [comment]
Any updates on this? I might be underestimating the difficulty but it seems like the NodeInjector could easily implement the correct Injector function. It is currently passing in undefined to getOrCreateInjectable.

```
export class NodeInjector implements Injector {
  constructor(
      private _tNode: TElementNode|TContainerNode|TElementContainerNode|null,
      private _lView: LView) {}
  get(token: any, notFoundValue?: any): any {
    return getOrCreateInjectable(this._tNode, this._lView, token, undefined, notFoundValue);
  }
}
```
